File: Sources/TransferMonitor.swift
import SwiftUI

/// Polls rclone's RC API for live transfer stats.
@MainActor
class TransferMonitor: ObservableObject {
    static let shared = TransferMonitor()

    @Published var transfers: [TransferInfo] = []
    @Published var globalStats: GlobalStats = GlobalStats()

    private var rcPorts: [String: Int] = [:]  // remote name -> RC port
    private var pollTimer: Timer?
    private var isPolling = false
    private var emptyPollCount = 0  // consecutive polls with no active transfers

    struct TransferInfo: Identifiable {
        var id: String { name + remoteName }
        let remoteName: String
        let name: String
        let size: Int64
        let bytes: Int64
        let speed: Double
        let percentage: Int
        let isUpload: Bool

        var progress: Double {
            guard size > 0 else { return 0 }
            return Double(bytes) / Double(size)
        }

        var speedFormatted: String {
            formatBytes(Int64(speed)) + "/s"
        }

        var sizeFormatted: String {
            formatBytes(size)
        }

        var bytesFormatted: String {
            formatBytes(bytes)
        }
    }

    struct GlobalStats {
        var bytesTransferred: Int64 = 0
        var totalBytes: Int64 = 0
        var totalTransfers: Int = 0
        var completedTransfers: Int = 0
        var speed: Double = 0
        var errors: Int = 0
        var activeTransfers: Int = 0

        var speedFormatted: String {
            formatBytes(Int64(speed)) + "/s"
        }

        var progress: Double {
            guard totalBytes > 0 else { return 0 }
            return Double(bytesTransferred) / Double(totalBytes)
        }

        var progressFormatted: String {
            "\(formatBytes(bytesTransferred)) / \(formatBytes(totalBytes))"
        }
    }

    func registerRC(remoteName: String, port: Int) {
        rcPorts[remoteName] = port
        startPolling()
    }

    func unregisterRC(remoteName: String) {
        rcPorts.removeValue(forKey: remoteName)
        // Immediately remove transfers for this remote so UI doesn't show stale data
        transfers.removeAll { $0.remoteName == remoteName }
        if rcPorts.isEmpty {
            stopPolling()
            transfers = []
            globalStats = GlobalStats()
        }
    }

    func cancelTransfer(_ transfer: TransferInfo) async {
        guard let port = rcPorts[transfer.remoteName],
              let url = URL(string: "http://127.0.0.1:\(port)/vfs/forget") else { return }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try? JSONSerialization.data(withJSONObject: ["file": transfer.name])
        request.timeoutInterval = 2
        _ = try? await URLSession.shared.data(for: request)
        transfers.removeAll { $0.id == transfer.id }
    }

    func unregisterAll() {
        rcPorts.removeAll()
        stopPolling()
        transfers = []
        globalStats = GlobalStats()
    }

    private func startPolling() {
        guard pollTimer == nil else { return }
        pollTimer = Timer.scheduledTimer(withTimeInterval: 3.0, repeats: true) { [weak self] _ in
            Task { @MainActor in
                await self?.poll()
            }
        }
    }

    private func stopPolling() {
        pollTimer?.invalidate()
        pollTimer = nil
    }

    private func poll() async {
        guard !isPolling else { return }
        isPolling = true
        defer { isPolling = false }
        var allTransfers: [TransferInfo] = []
        var combinedStats = GlobalStats()

        for (remoteName, port) in rcPorts {
            guard let stats = await fetchStats(port: port) else { continue }

            combinedStats.bytesTransferred += stats.bytesTransferred
            combinedStats.totalBytes += stats.totalBytes
            combinedStats.totalTransfers += stats.totalTransfers
            combinedStats.completedTransfers += stats.completedTransfers
            combinedStats.speed += stats.speed
            combinedStats.errors += stats.errors
            combinedStats.activeTransfers += stats.activeTransfers

            for t in stats.transferring {
                allTransfers.append(TransferInfo(
                    remoteName: remoteName,
                    name: t.name,
                    size: t.size,
                    bytes: t.bytes,
                    speed: t.speed,
                    percentage: t.percentage,
                    isUpload: false // rclone RC doesn't expose transfer direction
                ))
            }
        }

        // Avoid flickering: only clear the transfer list after 3 consecutive
        // empty polls (~9s), since rclone momentarily reports no active transfers
        // between chunks.
        if allTransfers.isEmpty && !transfers.isEmpty {
            emptyPollCount += 1
            if emptyPollCount < 3 { return }
        } else {
            emptyPollCount = 0
        }

        transfers = allTransfers
        globalStats = combinedStats
    }

    private struct RCStats {
        var bytesTransferred: Int64 = 0
        var totalBytes: Int64 = 0
        var totalTransfers: Int = 0
        var completedTransfers: Int = 0
        var speed: Double = 0
        var errors: Int = 0
        var activeTransfers: Int = 0
        var transferring: [RCTransfer] = []
    }

    private struct RCTransfer {
        var name: String
        var size: Int64
        var bytes: Int64
        var speed: Double
        var percentage: Int
    }

    private func fetchStats(port: Int) async -> RCStats? {
        guard let url = URL(string: "http://127.0.0.1:\(port)/core/stats") else { return nil }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = 2

        do {
            let (data, _) = try await URLSession.shared.data(for: request)
            guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                return nil
            }

            var stats = RCStats()
            stats.bytesTransferred = (json["bytes"] as? NSNumber)?.int64Value ?? 0
            stats.totalBytes = (json["totalBytes"] as? NSNumber)?.int64Value ?? 0
            stats.totalTransfers = (json["totalTransfers"] as? Int) ?? 0
            stats.completedTransfers = (json["transfers"] as? Int) ?? 0
            stats.speed = (json["speed"] as? Double) ?? 0
            // rclone accumulates harmless "errors" (e.g. Finder probing .DS_Store) — ignore them
            stats.errors = 0

            if let transferring = json["transferring"] as? [[String: Any]] {
                stats.activeTransfers = transferring.count
                stats.transferring = transferring.map { t in
                    RCTransfer(
                        name: (t["name"] as? String) ?? "unknown",
                        size: (t["size"] as? NSNumber)?.int64Value ?? 0,
                        bytes: (t["bytes"] as? NSNumber)?.int64Value ?? 0,
                        speed: (t["speed"] as? Double) ?? (t["speedAvg"] as? Double) ?? 0,
                        percentage: (t["percentage"] as? Int) ?? 0
                    )
                }
            }

            return stats
        } catch {
            return nil
        }
    }
}

// MARK: - Transfer Activity View

struct TransferActivityView: View {
    @StateObject private var monitor = TransferMonitor.shared
    @State private var transfers: [TransferMonitor.TransferInfo] = []
    @State private var stats: TransferMonitor.GlobalStats = .init()

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            if transfers.isEmpty && stats.activeTransfers == 0 {
                HStack {
                    Image(systemName: "checkmark.circle")
                        .foregroundStyle(.green)
                    Text("No active transfers")
                        .font(.caption)
                        .foregroundStyle(.secondary)
                }
                .padding(.vertical, 4)
            } else {
                // Global progress
                VStack(alignment: .leading, spacing: 6) {
                    HStack(spacing: 12) {
                        Label("\(stats.activeTransfers) active", systemImage: "arrow.triangle.2.circlepath")
                        Spacer()
                        Text(stats.speedFormatted)
                            .monospacedDigit()
                        if stats.errors > 0 {
                            Text("\(stats.errors) errors")
                                .foregroundStyle(.red)
                        }
                    }
                    .font(.caption2)
                    .foregroundStyle(.secondary)

                    if stats.totalBytes > 0 {
                        ProgressView(value: stats.progress)
                            .tint(.blue)
                        HStack {
                            Text(stats.progressFormatted)
                                .font(.caption2)
                                .foregroundStyle(.tertiary)
                            Spacer()
                            Text("\(stats.completedTransfers)/\(stats.totalTransfers)")
                                .font(.caption2)
                                .foregroundStyle(.tertiary)
                        }
                    }
                }

                // Individual transfers
                ScrollView {
                    VStack(spacing: 4) {
                        ForEach(transfers) { transfer in
                            HStack(spacing: 8) {
                                Image(systemName: "arrow.up.arrow.down.circle.fill")
                                    .foregroundStyle(.blue)
                                    .font(.caption)
                                VStack(alignment: .leading, spacing: 2) {
                                    Text(transfer.name)
                                        .font(.caption)
                                        .lineLimit(1)
                                        .truncationMode(.middle)
                                    if transfer.size > 0 {
                                        Text(transfer.sizeFormatted)
                                            .font(.caption2)
                                            .foregroundStyle(.tertiary)
                                    }
                                }
                                Spacer()
                                if transfer.speed > 0 {
                                    Text(transfer.speedFormatted)
                                        .font(.caption2)
                                        .monospacedDigit()
                                        .foregroundStyle(.secondary)
                                }
                            }
                            .padding(8)
                            .background(Color.primary.opacity(0.03), in: RoundedRectangle(cornerRadius: 6))
                        }
                    }
                }
                .frame(maxHeight: 250)
            }
        }
        .onAppear {
            transfers = monitor.transfers
            stats = monitor.globalStats
        }
        .onReceive(monitor.objectWillChange) { _ in
            DispatchQueue.main.async {
                transfers = monitor.transfers
                stats = monitor.globalStats
            }
        }
    }
}

// MARK: - Byte Formatting

func formatBytes(_ bytes: Int64) -> String {
    let units = ["B", "KB", "MB", "GB", "TB"]
    var value = Double(bytes)
    var unitIndex = 0
    while value >= 1024 && unitIndex < units.count - 1 {
        value /= 1024
        unitIndex += 1
    }
    if unitIndex == 0 { return "\(bytes) B" }
    return String(format: "%.1f %@", value, units[unitIndex])
}

File: Windows/SkyHook/Services/TransferMonitor.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using System.Net.Http;
using System.Text.Json;
using System.Threading.Tasks;
using System.Windows.Threading;
using CommunityToolkit.Mvvm.ComponentModel;

namespace SkyHook.Services;

/// <summary>
/// Polls rclone's RC API for live transfer stats.
/// Ported from macOS TransferMonitor.swift.
/// Singleton with ObservableObject. Uses DispatcherTimer for periodic polling.
/// </summary>
public partial class TransferMonitor : ObservableObject
{
    private static readonly Lazy<TransferMonitor> _lazy = new(() => new TransferMonitor());
    public static TransferMonitor Instance => _lazy.Value;

    // ── Observable Properties ────────────────────────────────────

    [ObservableProperty]
    private ObservableCollection<TransferInfo> _transfers = new();

    [ObservableProperty]
    private GlobalStats _combinedStats = new();

    // ── Private State ────────────────────────────────────────────

    private readonly Dictionary<string, int> _rcPorts = new(); // remote name -> RC port
    private DispatcherTimer? _pollTimer;
    private bool _isPolling;
    private int _emptyPollCount; // consecutive polls with no active transfers
    private readonly HttpClient _httpClient = new() { Timeout = TimeSpan.FromSeconds(2) };

    private TransferMonitor() { }

    // ── Public API ───────────────────────────────────────────────

    public void RegisterRC(string remoteName, int port)
    {
        _rcPorts[remoteName] = port;
        StartPolling();
    }

    public void UnregisterRC(string remoteName)
    {
        _rcPorts.Remove(remoteName);

        // Immediately remove transfers for this remote so UI doesn't show stale data
        var remaining = Transfers.Where(t => t.RemoteName != remoteName).ToList();
        Transfers = new ObservableCollection<TransferInfo>(remaining);

        if (_rcPorts.Count == 0)
        {
            StopPolling();
            Transfers = new ObservableCollection<TransferInfo>();
            CombinedStats = new GlobalStats();
        }
    }

    public void UnregisterAll()
    {
        _rcPorts.Clear();
        StopPolling();
        Transfers = new ObservableCollection<TransferInfo>();
        CombinedStats = new GlobalStats();
    }

    // ── Polling ──────────────────────────────────────────────────

    private void StartPolling()
    {
        if (_pollTimer != null) return;

        _pollTimer = new DispatcherTimer
        {
            Interval = TimeSpan.FromSeconds(3)
        };
        _pollTimer.Tick += async (_, _) => await Poll();
        _pollTimer.Start();
    }

    private void StopPolling()
    {
        _pollTimer?.Stop();
        _pollTimer = null;
    }

    private async Task Poll()
    {
        if (_isPolling) return;
        _isPolling = true;

        try
        {
            var allTransfers = new List<TransferInfo>();
            var combined = new GlobalStats();

            foreach (var (remoteName, port) in _rcPorts.ToList())
            {
                var stats = await FetchStats(port);
                if (stats == null) continue;

                combined.BytesTransferred += stats.BytesTransferred;
                combined.TotalBytes += stats.TotalBytes;
                combined.TotalTransfers += stats.TotalTransfers;
                combined.CompletedTransfers += stats.CompletedTransfers;
                combined.Speed += stats.Speed;
                // rclone accumulates harmless errors (e.g. Explorer probing thumbs.db) - ignore
                combined.Errors += 0;
                combined.ActiveTransfers += stats.ActiveTransfers;

                foreach (var t in stats.Transferring)
                {
                    allTransfers.Add(new TransferInfo
                    {
                        RemoteName = remoteName,
                        Name = t.Name,
                        Size = t.Size,
                        Bytes = t.Bytes,
                        Speed = t.Speed,
                        Percentage = t.Percentage
                    });
                }
            }

            // Avoid flickering: only clear the transfer list after 3 consecutive
            // empty polls (~9s), since rclone momentarily reports no active transfers
            // between chunks.
            if (allTransfers.Count == 0 && Transfers.Count > 0)
            {
                _emptyPollCount++;
                if (_emptyPollCount < 3) return;
            }
            else
            {
                _emptyPollCount = 0;
            }

            Transfers = new ObservableCollection<TransferInfo>(allTransfers);
            CombinedStats = combined;
        }
        finally
        {
            _isPolling = false;
        }
    }

    // ── RC Stats Fetch ───────────────────────────────────────────

    private async Task<RCStats?> FetchStats(int port)
    {
        try
        {
            var request = new HttpRequestMessage(HttpMethod.Post, $"http://127.0.0.1:{port}/core/stats");
            var response = await _httpClient.SendAsync(request);
            var json = await response.Content.ReadAsStringAsync();

            using var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;

            var stats = new RCStats
            {
                BytesTransferred = root.TryGetProperty("bytes", out var bytes) ? bytes.GetInt64() : 0,
                TotalBytes = root.TryGetProperty("totalBytes", out var totalBytes) ? totalBytes.GetInt64() : 0,
                TotalTransfers = root.TryGetProperty("totalTransfers", out var totalTransfers) ? totalTransfers.GetInt32() : 0,
                CompletedTransfers = root.TryGetProperty("transfers", out var transfers) ? transfers.GetInt32() : 0,
                Speed = root.TryGetProperty("speed", out var speed) ? speed.GetDouble() : 0,
                Errors = 0
            };

            if (root.TryGetProperty("transferring", out var transferring) &&
                transferring.ValueKind == JsonValueKind.Array)
            {
                stats.ActiveTransfers = transferring.GetArrayLength();
                foreach (var t in transferring.EnumerateArray())
                {
                    stats.Transferring.Add(new RCTransfer
                    {
                        Name = t.TryGetProperty("name", out var n) ? n.GetString() ?? "unknown" : "unknown",
                        Size = t.TryGetProperty("size", out var s) ? s.GetInt64() : 0,
                        Bytes = t.TryGetProperty("bytes", out var b) ? b.GetInt64() : 0,
                        Speed = t.TryGetProperty("speed", out var sp) ? sp.GetDouble() :
                                t.TryGetProperty("speedAvg", out var sa) ? sa.GetDouble() : 0,
                        Percentage = t.TryGetProperty("percentage", out var p) ? p.GetInt32() : 0
                    });
                }
            }

            return stats;
        }
        catch
        {
            return null;
        }
    }

    // ── Internal Models ──────────────────────────────────────────

    private class RCStats
    {
        public long BytesTransferred { get; set; }
        public long TotalBytes { get; set; }
        public int TotalTransfers { get; set; }
        public int CompletedTransfers { get; set; }
        public double Speed { get; set; }
        public int Errors { get; set; }
        public int ActiveTransfers { get; set; }
        public List<RCTransfer> Transferring { get; set; } = new();
    }

    private class RCTransfer
    {
        public string Name { get; set; } = string.Empty;
        public long Size { get; set; }
        public long Bytes { get; set; }
        public double Speed { get; set; }
        public int Percentage { get; set; }
    }
}

// ── Public Data Classes ──────────────────────────────────────

/// <summary>
/// Represents a single active file transfer.
/// </summary>
public class TransferInfo
{
    public string RemoteName { get; set; } = string.Empty;
    public string Name { get; set; } = string.Empty;
    public long Size { get; set; }
    public long Bytes { get; set; }
    public double Speed { get; set; }
    public int Percentage { get; set; }

    public string Id => $"{Name}{RemoteName}";

    public double Progress => Size > 0 ? (double)Bytes / Size : 0;

    public string SpeedFormatted => FormatBytes((long)Speed) + "/s";

    public string SizeFormatted => FormatBytes(Size);

    public string BytesFormatted => FormatBytes(Bytes);

    private static string FormatBytes(long bytes)
    {
        string[] units = { "B", "KB", "MB", "GB", "TB" };
        double value = bytes;
        int unitIndex = 0;
        while (value >= 1024 && unitIndex < units.Length - 1)
        {
            value /= 1024;
            unitIndex++;
        }
        if (unitIndex == 0) return $"{bytes} B";
        return $"{value:F1} {units[unitIndex]}";
    }
}

/// <summary>
/// Aggregated stats across all mounted remotes.
/// </summary>
public partial class GlobalStats : ObservableObject
{
    [ObservableProperty]
    private long _bytesTransferred;

    [ObservableProperty]
    private long _totalBytes;

    [ObservableProperty]
    private int _totalTransfers;

    [ObservableProperty]
    private int _completedTransfers;

    [ObservableProperty]
    private double _speed;

    [ObservableProperty]
    private int _errors;

    [ObservableProperty]
    private int _activeTransfers;

    public double Progress => TotalBytes > 0 ? (double)BytesTransferred / TotalBytes : 0;

    public string SpeedFormatted => FormatBytes((long)Speed) + "/s";

    public string ProgressFormatted => $"{FormatBytes(BytesTransferred)} / {FormatBytes(TotalBytes)}";

    private static string FormatBytes(long bytes)
    {
        string[] units = { "B", "KB", "MB", "GB", "TB" };
        double value = bytes;
        int unitIndex = 0;
        while (value >= 1024 && unitIndex < units.Length - 1)
        {
            value /= 1024;
            unitIndex++;
        }
        if (unitIndex == 0) return $"{bytes} B";
        return $"{value:F1} {units[unitIndex]}";
    }
}
